core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java

```diff
@@ -241,6 +241,20 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
         });
     }
 
+    /**
+     * {@code channel().remoteAddress()} can return {@code null} if the connection is closed
+     * before {@code channel().remoteAddress()} is called which caches the address in it.
+     */
+    private SocketAddress remoteAddress(ChannelHandlerContext ctx) {
+        SocketAddress remoteAddress = ctx.channel().remoteAddress();
+        if (remoteAddress == null) {
+            // This is called after connect() method is called so we have the remoteAddress.
+            assert this.remoteAddress != null;
+            remoteAddress = this.remoteAddress;
+        }
+        return remoteAddress;
```

Comment:
nit: How about simplifying this logic? 😉
```java
SocketAddress remoteAddress = ctx.channel().remoteAddress();
return firstNonNull(remoteAddress, this.remoteAddress);
```

Reply:
I’m not expert in this logic. I just wonder we don’t have to update `this.remoteAddress` just in case  `remoteAdress != null` and `remoteAdress != this.remoteAddress`.

Reply:
> return firstNonNull(remoteAddress, this.remoteAddress);

That's a good suggestion. I will take it. Thanks!

> I just wonder we don’t have to update this.remoteAddress = remoteAddress

I think we don't have to. The `remoteAddress` is only used later to extract host in `UpgradeRequestHandler`. We only need the host and the port, so I think we don't have to do that.